### PR TITLE
[Typing][B-99] Add type annotations for `python/paddle/distributed/parallel.py`

### PR DESCRIPTION
**Pull request by enkilee**
### PR Category

User Experience

### PR Types

Improvements

### Description

- https://github.com/PaddlePaddle/Paddle/issues/65008
B99

### CODE REVIEW

File: python/paddle/distributed/parallel.py [hunk 1]
```diff
@@ -126,7 +135,9 @@ def _split_tensors(coalesced_grads_and_grad_vars):
 
 @imperative_base.no_grad
 @framework.dygraph_only
-def build_groups(vars, group_size):
+def build_groups(
+    vars: Layer, group_size: int
+) -> list[list[Tensor, list[Tensor], list[int]]]:
```
Comment (megemini) [inline python/paddle/distributed/parallel.py]:
```suggestion
def build_groups(
    vars: list[Tensor], group_size: int
) -> list[list[Tensor | list[Tensor] | list[int]]]:
```

这里输出要用 `|` ～

File: python/paddle/distributed/parallel.py [hunk 2]
```diff
@@ -351,14 +362,21 @@ class DataParallel(layers.Layer):
...
-        last_comm_buffer_size=1,
-        find_unused_parameters=False,
-        group=None,
+        layers: Layer,
+        strategy: ParallelStrategy | None = None,
+        comm_buffer_size: int = 25,
+        last_comm_buffer_size: float = 1,
+        find_unused_parameters: bool = False,
+        group: Group | None = None,
     ):
```
Comment (megemini) [inline python/paddle/distributed/parallel.py]:
```suggestion
    ) -> None:
```

File: python/paddle/distributed/parallel.py [hunk 3]
```diff
@@ -351,14 +362,21 @@ class DataParallel(layers.Layer):
 
     """
 
+    find_unused_parameters: bool
+    grad_need_sync: bool
+    group: Group | None
+    var_dtype: core.VarDesc.VarType
```
Comment (megemini) [inline python/paddle/distributed/parallel.py]: `Tensor` ？

File: python/paddle/distributed/parallel.py [hunk 4]
```diff
@@ -495,7 +513,7 @@ def _find_tensor(self, obj):
         return []
 
     @contextmanager
-    def no_sync(self):
+    def no_sync(self) -> Generator[Any, Any, Any]:
```
Comment (megemini) [inline python/paddle/distributed/parallel.py]:
```suggestion
    def no_sync(self) -> Generator[None, None, None]:
```

File: python/paddle/distributed/parallel.py [hunk 5]
```diff
@@ -608,7 +626,9 @@ def state_dict(
         )
 
     @framework.deprecate_stat_dict
-    def set_state_dict(self, state_dict, use_structured_name=True):
+    def set_state_dict(
+        self, state_dict: _StateDict, use_structured_name: bool = True
+    ) -> tuple[list[str], list[str]]:
```
Comment (megemini) [inline python/paddle/distributed/parallel.py]:
```suggestion
    ) -> None:
```

File: python/paddle/distributed/parallel.py [hunk 6]
```diff
@@ -805,7 +825,7 @@ def device_type(self):
         return self._device_type
 
     @property
-    def current_endpoint(self):
+    def current_endpoint(self) -> list[str]:
```
Comment (megemini) [inline python/paddle/distributed/parallel.py]:
```suggestion
    def current_endpoint(self) -> str:
```

Comment (enkilee) [reply python/paddle/distributed/parallel.py]: 收到

File: python/paddle/distributed/parallel.py (hunk 5 above)
Comment (enkilee) [reply python/paddle/distributed/parallel.py]: 收到

File: python/paddle/distributed/parallel.py (hunk 4 above)
Comment (enkilee) [reply python/paddle/distributed/parallel.py]: 收到

File: python/paddle/distributed/parallel.py (hunk 3 above)
Comment (enkilee) [reply python/paddle/distributed/parallel.py]: 收到

File: python/paddle/distributed/parallel.py (hunk 2 above)
Comment (enkilee) [reply python/paddle/distributed/parallel.py]: 收到

File: python/paddle/distributed/parallel.py (hunk 1 above)
Comment (enkilee) [reply python/paddle/distributed/parallel.py]: 收到

**Review by SigureMo — APPROVED** [review]
LGTMeow

**Pull request merged**